File: NumericalDifferentiationIntegration/diff.py
def f(x):
    return x**2 + 3*x - 4

def dfdx(x):
    return 2*x + 3

def forward_difference(f, x, dx=0.01):
    """Forward difference scheme with appoximation error O(h)"""
    return (f(x + dx) - f(x)) / dx

def backward_difference(f, x, dx=0.01):
    """Backward difference scheme with appoximation error O(h)"""
    return (f(x) - f(x - dx)) / dx

def central_difference(f, x, dx=0.01):
    """Central difference scheme with appoximation error O(h^2)"""
    return (f(x + dx) - f(x - dx)) / (2*dx)

def central_h4(f, x, dx=0.01):
    """Centreped formula with appoximation error O(h^4)"""
    return (-f(x + 2*dx) + 8*f(x + dx) - 8*f(x - dx) + f(x - 2*dx)) / (12*dx)

def error_evaluation(F, N, f, x, dx=0.01):
    print(N.__doc__)
    exact = F(x)
    numeric = N(f, x, dx)
    error = exact - numeric
    print(f"Exact: {exact}")
    print(f"Numeric: {numeric}")
    print(f"Absolute error: {abs(error)}")


x = float(input("x = "))

error_evaluation(dfdx, forward_difference, f, x)
error_evaluation(dfdx, backward_difference, f, x)
error_evaluation(dfdx, central_difference, f, x)
error_evaluation(dfdx, central_h4, f, x)

File: NumericalDifferentiationIntegration/diff.cpp
#include <iostream>
#include <cmath>

double f(double x) {
    return x*x + 3*x - 4;
}

double dfdx(double x) {
    return 2*x + 3;
}

double forward_difference(double x, double dx=0.01) {
    // Forward difference scheme with appoximation error O(h)
    return (f(x + dx) - f(x)) / dx;
}

double backward_difference(double x, double dx=0.01) {
    // Backward difference scheme with appoximation error O(h)
    return (f(x) - f(x - dx)) / dx;
}

double central_difference(double x, double dx=0.01) {
    // Central difference scheme with appoximation error O(h^2)
    return (f(x + dx) - f(x - dx)) / (2*dx);
}

void error_evaluation(double (*f)(double x, double dx), 
                        double x, double dx=0.01, std::string name="diff") {
    double exact = dfdx(x);
    double numeric = f(x, dx);
    double error = exact - numeric;

    printf("Method: %s\n", name.c_str());
    printf("Exact: %e\n", exact);
    printf("Numeric: %e\n", numeric);
    printf("Absolute error: %e\n", fabs(error));
}


int main() {
    double x, dx;
    std::cout << "x = ";
    std::cin >> x;
    std::cout << "dx = ";
    std::cin >> dx;
    error_evaluation(forward_difference, x, dx, "Forward O(h)");
    error_evaluation(backward_difference, x, dx, "Backward O(h)");
    error_evaluation(central_difference, x, dx, "Central O(h^2)");

    return 0;
}
